### DISCUSSION TITLE
Issues building MicroPython 1.20.0 for Windows: no native or viper code emitters in MSVC, MINGW/MSYS2 ?

### DISCUSSION THREAD

**Discussion by AltoRetrato** [micropython/micropython#12333] [Unix & Windows]
Hello, MicroPython World!

I want to build MicroPython for Windows with support for native & viper code emitters. Is it possible? In my case, there were issues with missing `mp_unix_mark_exec`.

### [Building using MS Visual Studio 2013 (or higher)](https://github.com/micropython/micropython/tree/master/ports/windows#building-using-ms-visual-studio-2013-or-higher)
I built `micropython.exe` with VS2019, but when I try to run a function using `@micropython.native` or `@micropython.viper`, e.g.:
```py
@micropython.viper
def add_ints():
    return 1 + 2
```
... I get a `SyntaxError: invalid micropython decorator`.

If I set `#define MICROPY_EMIT_X64  (1)` in `mpconfigport.h`, building the solution with VS2019 fails with:
```
1>C:\micropython\py\objgenerator.c(213,1): error C2220: the following warning is treated as an error
1>C:\micropython\py\objgenerator.c(213,1): warning C4090: 'initializing': different 'const' qualifiers
```

Relevant code:
```c
209   #if MICROPY_EMIT_NATIVE
210    if (self->code_state.exc_sp_idx == MP_CODE_STATE_EXC_SP_IDX_SENTINEL) {
211        // A native generator, with entry point 2 words into the "bytecode" pointer
212        typedef uintptr_t (*mp_fun_native_gen_t)(void *, mp_obj_t);
213        mp_fun_native_gen_t fun = MICROPY_MAKE_POINTER_CALLABLE((const void *)(self->code_state.fun_bc->bytecode + 2 * sizeof(uintptr_t)));
214        ret_kind = fun((void *)&self->code_state, throw_value);
215    } else
```

I tried changing `typedef uintptr_t` to `typedef const uintptr_t`, but then build fails with:
```
1>gccollect.obj : error LNK2019: unresolved external symbol _mp_unix_mark_exec referenced in function _gc_collect
1>C:\micropython\ports\windows\build-standard\ReleaseWin32\micropython.exe : fatal error LNK1120: 1 unresolved externals
```
Then I gave up on MSVC and tried...

### [Building under MSYS2](https://github.com/micropython/micropython/tree/master/ports/windows#building-under-msys2)
Again, I can build MicroPython with MSYS2, except when I set `#define MICROPY_EMIT_X64  (1)` in `mpconfigport.h`

```
$ make
[...]
CC ../../ports/unix/gccollect.c
../../ports/unix/gccollect.c: In function 'gc_collect':
../../ports/unix/gccollect.c:43:5: error: implicit declaration of function 'mp_unix_mark_exec' [-Werror=implicit-function-declaration]
   43 |     mp_unix_mark_exec();
      |     ^~~~~~~~~~~~~~~~~
cc1.exe: all warnings being treated as errors
See https://github.com/micropython/micropython/wiki/Build-Troubleshooting
make: *** [../../py/mkrules.mk:83: build-standard/ports/unix/gccollect.o] Error 1
```
The [Build-Troubleshooting](https://github.com/micropython/micropython/wiki/Build-Troubleshooting) page didn't help.

I noticed `void mp_unix_mark_exec(void);` is declared in `ports/unix/mpconfigport.h` and implemented in `ports/unix/alloc.c`... so I'm not sure if those files are not being correctly processed or if there is no support for code emitters on Windows.

Could someone clarify this?

Thank you!

**Comment by dlech** [comment]
> 1>gccollect.obj : error LNK2019: unresolved external symbol _mp_unix_mark_exec referenced in function _gc_collect
> 1>C:\micropython\ports\windows\build-standard\ReleaseWin32\micropython.exe : fatal error LNK1120: 1 unresolved externals

This sounds like alloc.c is missing from the list of source files.

> ../../ports/unix/gccollect.c:43:5: error: implicit declaration of function 'mp_unix_mark_exec' [-Werror=implicit-function-declaration]

Sounds like this file is missing an include of mpconfig.h

**Reply by AltoRetrato** [reply]
> Sounds like this file is missing an include of mpconfig.h

If I understood correctly, there should be only one `mpconfigport.h` per port, and including it in the file would cause errors. I tried to `#include "mpconfigport.h"` in `ports/unix/gccollect.c` anyway, and there were several `redefined` errors.

I changed `ports\windows\mpconfigport.h` to include:
```c
void mp_unix_mark_exec(void);
```
... and built again with MSYS2, then got to the same point I got with MSVC:
```
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: build-standard/ports/unix/gccollect.o:gccollect.c:(.text+0xf): undefined reference to `mp_unix_mark_exec'
collect2.exe: error: ld returned 1 exit status
make: *** [../../py/mkrules.mk:211: build-standard/micropython.exe] Error 1
```

`ports/unix/alloc.c` includes `<sys/mman.h>`, which is not available on Windows, so I patched it to use `VirtualAlloc` instead for a quick test. I included `"alloc.c"` in `ports/unix/gccollect.c` and could build without errors.

I can now declare a function with an emitter, but unsurprisingly, it crashes when I invoke it:
```
>>> @micropython.viper
... def add_ints() -> int:
...     return 1 + 2
...
>>> add_ints()
Segmentation fault
```
So I have the impression that emitters are not currently officially supported on Windows, and I'm not sure if there is a simple solution / workaround...

**Comment by andrewleech** [comment]
I've run into a similar issue in the past, documented in https://github.com/micropython/micropython/issues/7894

It also links to more older posts.
